### PR TITLE
Fix for TIKA-2262: Supporting Image-to-Text (Image Captioning) in Tika

### PR DESCRIPTION
**Pull request by ThejanW**
TIKA 2262: Supporting Image-to-Text (Image Captioning) in Tika for Image MIME Types

**Image captioning server - Complete**
_How to test the server?_
**(Assuming that you've already built the image inception-rest-tika in tika-parsers/src/main/resources/org/apache/tika/parser/recognition/tf/)** 

1. Build the dockerfile with this command,
**"docker build -f Im2txtRestDockerfile -t im2txt-rest-tika ."**
2. Run the container with this command, 
**"docker run -it -p 8764:8764 im2txt-rest-tika"**
3. Tryout this test url,
http://localhost:8764/inception/v3/captions?url=http://img09.deviantart.net/5281/i/2009/326/a/f/girl_walk_in_the_water_by_jmd4040116.jpg&beam_size=3&max_caption_length=15
4. Change the post parameters url, beam_size & max_caption_length as you prefer & test further.

**Java client - Complete**
_How to test the client?_

1. When the server is running, run the jpegRESTim2txtTest() in ObjectRecognitionParserTest class
2. Change the parameters in config.xml file & test further.

### CODE REVIEW

**Comment by ThejanW** [comment]
@thammegowda @chrismattmann ping!

**Comment by ThejanW** [comment]
@tballison my tests are failing. Any ideas on what's causing these errors?

**Comment by ThejanW** [comment]
Error:Error:line (26)java: cannot find symbol
  symbol:   class MemoryLimitException
  location: package org.apache.commons.compress
    Error:Error:line (120)java: cannot find symbol
  symbol:   variable BROTLI
  location: class org.apache.commons.compress.compressors.CompressorStreamFactory
    Error:Error:line (122)java: cannot find symbol
  symbol:   variable LZ4_BLOCK
  location: class org.apache.commons.compress.compressors.CompressorStreamFactory
    Error:Error:line (124)java: cannot find symbol
  symbol:   variable LZ4_FRAMED
  location: class org.apache.commons.compress.compressors.CompressorStreamFactory
    Error:Error:line (177)java: no suitable constructor found for CompressorStreamFactory(boolean,int)
    constructor org.apache.commons.compress.compressors.CompressorStreamFactory.CompressorStreamFactory() is not applicable
      (actual and formal argument lists differ in length)
    constructor org.apache.commons.compress.compressors.CompressorStreamFactory.CompressorStreamFactory(boolean) is not applicable
      (actual and formal argument lists differ in length)
    Error:Error:line (180)java: cannot find symbol
  symbol:   class MemoryLimitException
  location: class org.apache.tika.parser.pkg.CompressorParser

**Comment by ThejanW** [comment]
/home/thejan/IdeaProjects/GSoC/tika/tika-parsers/src/main/java/org/apache/tika/parser/pkg/CompressorParser.java
    Error:Error:line (26)java: cannot find symbol
  symbol:   class MemoryLimitException
  location: package org.apache.commons.compress
    Error:Error:line (120)java: cannot find symbol
  symbol:   variable BROTLI
  location: class org.apache.commons.compress.compressors.CompressorStreamFactory
    Error:Error:line (122)java: cannot find symbol
  symbol:   variable LZ4_BLOCK
  location: class org.apache.commons.compress.compressors.CompressorStreamFactory
    Error:Error:line (124)java: cannot find symbol
  symbol:   variable LZ4_FRAMED
  location: class org.apache.commons.compress.compressors.CompressorStreamFactory
    Error:Error:line (177)java: no suitable constructor found for CompressorStreamFactory(boolean,int)
    constructor org.apache.commons.compress.compressors.CompressorStreamFactory.CompressorStreamFactory() is not applicable
      (actual and formal argument lists differ in length)
    constructor org.apache.commons.compress.compressors.CompressorStreamFactory.CompressorStreamFactory(boolean) is not applicable
      (actual and formal argument lists differ in length)
    Error:Error:line (180)java: cannot find symbol
  symbol:   class MemoryLimitException
  location: class org.apache.tika.parser.pkg.CompressorParser

============================================================================

/home/thejan/IdeaProjects/GSoC/tika/tika-parsers/src/main/java/org/apache/tika/parser/captioning/tf/TensorflowRESTCaptioner.java
    Error:Error:line (63)java: org.apache.tika.parser.captioning.tf.TensorflowRESTCaptioner is not abstract and does not override abstract method checkInitialization(org.apache.tika.config.InitializableProblemHandler) in org.apache.tika.config.Initializable

============================================================================

/home/thejan/IdeaProjects/GSoC/tika/tika-parsers/src/main/java/org/apache/tika/parser/mail/RFC822Parser.java
    Error:Error:line (63)java: cannot find symbol
  symbol:   class Builder
  location: class org.apache.james.mime4j.stream.MimeConfig

============================================================================

/home/thejan/IdeaProjects/GSoC/tika/tika-parsers/src/main/java/org/apache/tika/parser/pkg/ZipContainerDetector.java
    Error:Error:line (105)java: cannot find symbol
  symbol:   method detect(java.io.ByteArrayInputStream)
  location: class org.apache.commons.compress.compressors.CompressorStreamFactory
    Error:Error:line (114)java: cannot find symbol
  symbol:   method detect(java.io.ByteArrayInputStream)
  location: class org.apache.commons.compress.archivers.ArchiveStreamFactory

============================================================================

**Comment by ThejanW** [comment]
Object Recognition Parser, Video Labeling Parser don't work anymore.

**Comment by tballison** [comment]
I'm getting a clean build locally, and when I do a fresh clone on a different machine.  Those errors reflect changes in code without corresponding updates to dependencies.

**Comment by ThejanW** [comment]
@tballison I could fix the maven import errors, then there was error `TensorflowRESTCaptioner is not abstract and does not override abstract method checkInitialization(org.apache.tika.config.InitializableProblemHandler) in org.apache.tika.config.Initializable`

I had to put this in TensorflowRESTCaptioner to get rid of the error.
`@Override
    public void checkInitialization(InitializableProblemHandler handler) throws TikaConfigException {
}`

**Comment by ThejanW** [comment]
@thammegowda @chrismattmann please merge my commits so I can proceed.

**Comment by chrismattmann** [comment]
merged into the branch. I am going to test this right now. Let's keep working on the branch.

File: tika-parsers/src/main/resources/org/apache/tika/parser/captioning/tf/Im2txtRestDockerfile [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+MAINTAINER Apache Tika Team
+
+#Install python pillow. TODO: Remove this with the fix for TIKA 2398
+RUN pip install pillow
+
+# Download the pretrained im2txt checkpoint
+WORKDIR /usr/share/apache-tika/models/dl/image/caption/
+
+RUN \
+    wget https://www.dropbox.com/s/l9ignjpjk774n2z/model_main_untuned.zip?dl=0 \
```
Comment (chrismattmann) [inline tika-parsers/src/main/resources/org/apache/tika/parser/captioning/tf/Im2txtRestDockerfile]: can we please check in the model to https://github.com/USCDataScience/img2text.git? I can create the repo for you. Then please check in any scripts you used to generate the model. Then we can check in the model zip file.

**Comment by chrismattmann** [comment]
OK I was able to test this out. I fixed the unit tests to properly use junit.Assume. I also fixed a Locale issue with forbidden APIs. Got it working both unit test wise, and actually via java, both are shown below. I will commit this to 1.17-master tomorrow. GREAT WORK @ThejanW and @thammegowda 

## Unit tests

```
[INFO] --- maven-surefire-plugin:2.18.1:test (default-test) @ tika-parsers ---
[INFO] Surefire report directory: /Users/mattmann/src/tika/tika-parsers/target/surefire-reports

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.apache.tika.parser.recognition.ObjectRecognitionParserTest
log4j: reset attribute= "false".
log4j: Threshold ="null".
log4j: Retreiving an instance of org.apache.log4j.Logger.
log4j: Setting [ProgressAppender] additivity to [false].
log4j: Level value for ProgressAppender is  [INFO].
log4j: ProgressAppender level set to INFO
log4j: Class name: [org.apache.log4j.ConsoleAppender]
log4j: Parsing layout of class: "org.apache.log4j.PatternLayout"
log4j: Setting property [conversionPattern] to [%m].
log4j: Adding appender named [noEolAppender] to category [ProgressAppender].
log4j: Retreiving an instance of org.apache.log4j.Logger.
log4j: Setting [ProgressDone] additivity to [false].
log4j: Level value for ProgressDone is  [INFO].
log4j: ProgressDone level set to INFO
log4j: Class name: [org.apache.log4j.ConsoleAppender]
log4j: Parsing layout of class: "org.apache.log4j.PatternLayout"
log4j: Setting property [conversionPattern] to [%m%n].
log4j: Adding appender named [eolAppender] to category [ProgressDone].
log4j: Level value for root is  [INFO].
log4j: root level set to INFO
log4j: Class name: [org.apache.log4j.ConsoleAppender]
log4j: Parsing layout of class: "org.apache.log4j.PatternLayout"
log4j: Setting property [conversionPattern] to [%d{dd MMM yyyy HH:mm:ss} %5p %c{1} - %m%n].
log4j: Adding appender named [consoleAppender] to category [root].
09 Jul 2017 06:03:59  INFO TensorflowRESTCaptioner - Available = true, API Status = HTTP/1.0 200 OK
09 Jul 2017 06:03:59  INFO TensorflowRESTCaptioner - Captions = 5, MaxCaptionLength = 15
09 Jul 2017 06:03:59  INFO ObjectRecognitionParser - Recogniser = org.apache.tika.parser.captioning.tf.TensorflowRESTCaptioner
09 Jul 2017 06:03:59  INFO ObjectRecognitionParser - Recogniser Available = true
09 Jul 2017 06:03:59  INFO ObjectRecognitionParser - minConfidence = 0.05, topN=2
09 Jul 2017 06:04:01  INFO ObjectRecognitionParser - Time taken 1477ms
09 Jul 2017 06:04:01  INFO ObjectRecognitionParserTest - MetaValues = a baseball player holding a bat on a field . (0.00168) a baseball player holding a baseball bat on a field . (0.00163) a man in a baseball uniform throwing a baseball . (0.00090) a man in a baseball uniform holding a bat . (0.00068) a man in a baseball uniform holding a baseball bat . (0.00042)
09 Jul 2017 06:04:01  INFO TensorflowRESTCaptioner - Available = true, API Status = HTTP/1.0 200 OK
09 Jul 2017 06:04:01  INFO TensorflowRESTCaptioner - Captions = 5, MaxCaptionLength = 15
09 Jul 2017 06:04:01  INFO ObjectRecognitionParser - Recogniser = org.apache.tika.parser.captioning.tf.TensorflowRESTCaptioner
09 Jul 2017 06:04:01  INFO ObjectRecognitionParser - Recogniser Available = true
09 Jul 2017 06:04:01  INFO ObjectRecognitionParser - minConfidence = 0.05, topN=2
09 Jul 2017 06:04:02  INFO ObjectRecognitionParser - Time taken 1389ms
09 Jul 2017 06:04:02  INFO ObjectRecognitionParserTest - MetaValues = a baseball player pitching a ball on top of a field . (0.00367) a baseball player holding a bat on top of a field . (0.00272) a baseball player swinging a bat at a ball (0.00220) a baseball player holding a bat on a field . (0.00193) a baseball player is getting ready to hit a ball . (0.00146)
09 Jul 2017 06:04:02  INFO TensorflowRESTCaptioner - Available = true, API Status = HTTP/1.0 200 OK
09 Jul 2017 06:04:02  INFO TensorflowRESTCaptioner - Captions = 5, MaxCaptionLength = 15
09 Jul 2017 06:04:02  INFO ObjectRecognitionParser - Recogniser = org.apache.tika.parser.captioning.tf.TensorflowRESTCaptioner
09 Jul 2017 06:04:02  INFO ObjectRecognitionParser - Recogniser Available = true
09 Jul 2017 06:04:02  INFO ObjectRecognitionParser - minConfidence = 0.05, topN=2
09 Jul 2017 06:04:03  INFO ObjectRecognitionParser - Time taken 1500ms
09 Jul 2017 06:04:03  INFO ObjectRecognitionParserTest - MetaValues = a baseball player holding a bat on a field . (0.00150) a baseball player holding a baseball bat on a field . (0.00138) a baseball player holding a bat on top of a field . (0.00090) a man in a baseball uniform holding a bat . (0.00063) a man in a baseball uniform holding a baseball bat . (0.00037)
Tests run: 5, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 5.384 sec - in org.apache.tika.parser.recognition.ObjectRecognitionParserTest

Results :

Tests run: 5, Failures: 0, Errors: 0, Skipped: 2

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 9.027 s
[INFO] Finished at: 2017-07-08T23:04:03-07:00
[INFO] Final Memory: 33M/968M
[INFO] ------------------------------------------------------------------------
LMC-053601:tika-parsers mattmann$ 
```

## Java

```
LMC-053601:tika mattmann$ java -jar tika-app/target/tika-app-1.17-SNAPSHOT.jar --config=tika-parsers/src/test/resources/org/apache/tika/parser/recognition/tika-config-tflow-im2txt-rest.xml https://upload.wikimedia.org/wikipedia/commons/f/f6/Working_Dogs%2C_Handlers_Share_Special_Bond_DVIDS124942.jpg
INFO  Available = true, API Status = HTTP/1.0 200 OK
INFO  Captions = 5, MaxCaptionLength = 15
INFO  Recogniser = org.apache.tika.parser.captioning.tf.TensorflowRESTCaptioner
INFO  Recogniser Available = true
INFO  minConfidence = 0.05, topN=2
INFO  Time taken 1779ms
<?xml version="1.0" encoding="UTF-8"?><html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta name="org.apache.tika.parser.recognition.object.rec.impl" content="org.apache.tika.parser.captioning.tf.TensorflowRESTCaptioner"/>
<meta name="X-Parsed-By" content="org.apache.tika.parser.CompositeParser"/>
<meta name="X-Parsed-By" content="org.apache.tika.parser.recognition.ObjectRecognitionParser"/>
<meta name="resourceName" content="Working_Dogs%2C_Handlers_Share_Special_Bond_DVIDS124942.jpg"/>
<meta name="Content-Length" content="295937"/>
<meta name="CAPTION" content="a man standing next to a dog on a leash . (0.00017)"/>
<meta name="CAPTION" content="a man standing next to a dog on a bench . (0.00017)"/>
<meta name="CAPTION" content="a man and a dog are sitting on a bench . (0.00014)"/>
<meta name="CAPTION" content="a man and a dog sitting on a bench . (0.00013)"/>
<meta name="CAPTION" content="a man and a dog are sitting on a bench (0.00009)"/>
<meta name="Content-Type" content="image/jpeg"/>
<title/>
</head>
<body><ol id="captions">	<li id="0"> a man standing next to a dog on a leash . [en](confidence = 0.000167)</li>
	<li id="1"> a man standing next to a dog on a bench . [en](confidence = 0.000167)</li>
	<li id="2"> a man and a dog are sitting on a bench . [en](confidence = 0.000138)</li>
	<li id="3"> a man and a dog sitting on a bench . [en](confidence = 0.000131)</li>
	<li id="4"> a man and a dog are sitting on a bench [en](confidence = 0.000092)</li>
</ol>
</body></html>LMC-053601:tika mattmann$ 
```

Documentation is here: https://wiki.apache.org/tika/ImageCaption

**Pull request merged**